File: src/lib/supabaseClient.ts
/**
 * Supabase Client Configuration
 * Handles authentication and database connections
 */

import { createClient } from '@supabase/supabase-js';
import type {
  ChatHistory,
  MusicKnowledge,
  UserFeedback,
  CodetteFile,
} from '../types/supabase';

// Export operation groups types
interface ApiMetric {
  id: string;
  endpoint: string;
  method: string;
  response_time_ms: number;
  status_code: number;
  created_at?: string;
}

interface BenchmarkResult {
  id: string;
  benchmark_type: string;
  score: number;
  created_at?: string;
}

// Get Supabase credentials from environment
const SUPABASE_URL = import.meta.env.VITE_SUPABASE_URL || '';
const SUPABASE_ANON_KEY = import.meta.env.VITE_SUPABASE_ANON_KEY || '';

if (!SUPABASE_URL || !SUPABASE_ANON_KEY) {
  console.warn('⚠️ Supabase credentials not configured. Set VITE_SUPABASE_URL and VITE_SUPABASE_ANON_KEY');
}

// Initialize Supabase client
export const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Handle Supabase errors with consistent formatting
 */
export const handleSupabaseError = (error: any): string => {
  if (error?.message) return error.message;
  if (error?.error_description) return error.error_description;
  if (typeof error === 'string') return error;
  return 'An error occurred with the database';
};

/**
 * Execute Supabase query with error handling
 */
export const executeQuery = async <T,>(
  query: any
): Promise<{ data: T | null; error: string | null }> => {
  try {
    const { data, error } = await query;
    if (error) throw error;
    return { data, error: null };
  } catch (err) {
    const error = handleSupabaseError(err);
    console.error('Supabase query error:', error);
    return { data: null, error };
  }
};

// ============================================================================
// CHAT HISTORY OPERATIONS
// ============================================================================

export const chatHistoryOps = {
  /**
   * Get or create chat history for user
   */
  async getOrCreate(userId: string): Promise<{ data: ChatHistory | null; error: string | null }> {
    return executeQuery(
      supabase
        .from('chat_history')
        .select('*')
        .eq('user_id', userId)
        .single()
        .then(async (result) => {
          if (result.error?.code === 'PGRST116') {
            // No row found, create one
            return supabase.from('chat_history').insert({
              user_id: userId,
              messages: [],
            });
          }
          return result;
        })
    );
  },

  /**
   * Add message to chat history
   */
  async addMessage(userId: string, message: any) {
    const { data: chat, error: fetchError } = await supabase
      .from('chat_history')
      .select('messages')
      .eq('user_id', userId)
      .single();

    if (fetchError) return { data: null, error: handleSupabaseError(fetchError) };

    const messages = chat?.messages || [];
    messages.push(message);

    return executeQuery(
      supabase
        .from('chat_history')
        .update({ messages, updated_at: new Date().toISOString() })
        .eq('user_id', userId)
    );
  },

  /**
   * Clear chat history
   */
  async clear(userId: string) {
    return executeQuery(
      supabase
        .from('chat_history')
        .update({ messages: [], updated_at: new Date().toISOString() })
        .eq('user_id', userId)
    );
  },
};

// ============================================================================
// MUSIC KNOWLEDGE OPERATIONS
// ============================================================================

export const musicKnowledgeOps = {
  /**
   * Search music knowledge with similarity search using embeddings
   */
  async searchBySimilarity(
    embedding: number[],
    limit: number = 5,
    threshold: number = 0.7
  ): Promise<{ data: MusicKnowledge[] | null; error: string | null }> {
    return executeQuery(
      supabase.rpc('search_music_knowledge', {
        query_embedding: embedding,
        match_count: limit,
        similarity_threshold: threshold,
      })
    );
  },

  /**
   * Search music knowledge by text (full-text search)
   */
  async searchByText(query: string, limit: number = 10) {
    return executeQuery(
      supabase
        .from('music_knowledge')
        .select('*')
        .or(`fts.match.'${query}',topic.ilike.%${query}%,category.ilike.%${query}%`)
        .limit(limit)
    );
  },

  /**
   * Get music knowledge by category
   */
  async getByCategory(category: string, limit: number = 20) {
    return executeQuery(
      supabase
        .from('music_knowledge')
        .select('*')
        .eq('category', category)
        .order('confidence', { ascending: false })
        .limit(limit)
    );
  },

  /**
   * Add new music knowledge
   */
  async add(knowledge: Omit<MusicKnowledge, 'id' | 'created_at' | 'updated_at'>) {
    return executeQuery(supabase.from('music_knowledge').insert(knowledge));
  },

  /**
   * Update music knowledge
   */
  async update(id: string, updates: Partial<MusicKnowledge>) {
    return executeQuery(
      supabase
        .from('music_knowledge')
        .update({ ...updates, updated_at: new Date().toISOString() })
        .eq('id', id)
    );
  },
};

// ============================================================================
// CODETTE FILES OPERATIONS
// ============================================================================

export const codetteFilesOps = {
  /**
   * List all files by type
   */
  async getByType(fileType: string) {
    return executeQuery(
      supabase.from('codette_files').select('*').eq('file_type', fileType).order('uploaded_at', {
        ascending: false,
      })
    );
  },

  /**
   * Upload file metadata
   */
  async uploadMetadata(metadata: Omit<CodetteFile, 'id'>) {
    return executeQuery(supabase.from('codette_files').insert(metadata));
  },

  /**
   * Delete file metadata
   */
  async deleteMetadata(id: string) {
    return executeQuery(supabase.from('codette_files').delete().eq('id', id));
  },
};

// ============================================================================
// API METRICS OPERATIONS
// ============================================================================

export const apiMetricsOps = {
  /**
   * Log API request metrics
   */
  async logMetric(metric: Omit<ApiMetric, 'id' | 'created_at'>) {
    return executeQuery(supabase.from('api_metrics').insert(metric));
  },

  /**
   * Get average response time for endpoint
   */
  async getAverageResponseTime(endpoint: string, days: number = 7) {
    const since = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();
    return executeQuery(
      supabase.rpc('get_avg_response_time', {
        endpoint_path: endpoint,
        since_date: since,
      })
    );
  },
};

// ============================================================================
// BENCHMARK OPERATIONS
// ============================================================================

export const benchmarkOps = {
  /**
   * Record benchmark result
   */
  async recordResult(result: Omit<BenchmarkResult, 'id' | 'created_at'>) {
    return executeQuery(supabase.from('benchmark_results').insert(result));
  },

  /**
   * Get average score by test type
   */
  async getAverageScoreByType(testType: string) {
    return executeQuery(
      supabase.rpc('get_benchmark_average', {
        test_type_filter: testType,
      })
    );
  },

  /**
   * Get latest benchmarks
   */
  async getLatest(limit: number = 10) {
    return executeQuery(
      supabase
        .from('benchmark_results')
        .select('*')
        .order('created_at', { ascending: false })
        .limit(limit)
    );
  },
};

// ============================================================================
// FEEDBACK OPERATIONS
// ============================================================================

export const feedbackOps = {
  /**
   * Submit user feedback
   */
  async submit(feedback: Omit<UserFeedback, 'id' | 'created_at'>) {
    return executeQuery(supabase.from('user_feedback').insert(feedback));
  },

  /**
   * Get average rating
   */
  async getAverageRating(taskId?: string) {
    let query = supabase.from('user_feedback').select('rating');
    if (taskId) query = query.eq('task_id', taskId);
    return executeQuery(query);
  },
};

export default supabase;

File: daw_core/supabase_client.py
"""
Supabase Client Integration for Codette Backend
Handles all database operations and real-time subscriptions
"""

import os
from typing import Optional, Dict, Any, List, Tuple
from datetime import datetime
from dotenv import load_dotenv

try:
    from supabase import create_client, Client
    SUPABASE_AVAILABLE = True
except ImportError:
    SUPABASE_AVAILABLE = False
    print("[WARNING] supabase-py not installed. Install with: pip install supabase")

# Load environment variables
load_dotenv()

# ============================================================================
# SUPABASE CLIENT INITIALIZATION
# ============================================================================

SUPABASE_URL = os.getenv("SUPABASE_URL", "")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY", "")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY", "")

# Initialize Supabase client (use service role key for backend operations)
supabase: Optional[Client] = None

if SUPABASE_AVAILABLE and SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY:
    try:
        supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)
        print("[OK] Supabase client initialized successfully")
    except Exception as e:
        print(f"[WARNING] Could not initialize Supabase client: {e}")
        supabase = None
else:
    if not SUPABASE_AVAILABLE:
        print("[INFO] Supabase not available - backend will operate without database")
    elif not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
        print("[INFO] Supabase environment variables not configured - database disabled")


# ============================================================================
# ERROR HANDLING
# ============================================================================

class SupabaseError(Exception):
    """Custom exception for Supabase operations"""
    pass


def handle_supabase_error(error: Exception) -> Tuple[Dict[str, Any], int]:
    """Convert Supabase errors to HTTP responses"""
    error_str = str(error)

    if "duplicate key" in error_str.lower():
        return {"error": "Record already exists"}, 409
    elif "not found" in error_str.lower() or "no rows" in error_str.lower():
        return {"error": "Record not found"}, 404
    elif "permission denied" in error_str.lower():
        return {"error": "Access denied"}, 403
    elif "jwt" in error_str.lower() or "unauthorized" in error_str.lower():
        return {"error": "Unauthorized"}, 401
    else:
        return {"error": f"Database error: {error_str}"}, 500


# ============================================================================
# CHAT HISTORY OPERATIONS
# ============================================================================

async def get_or_create_chat_history(user_id: str) -> Tuple[Optional[Dict], Optional[str]]:
    """Get or create chat history for user"""
    if not supabase:
        return None, "Supabase not available"

    try:
        # Try to get existing chat history
        response = supabase.table("chat_history").select("*").eq("user_id", user_id).execute()

        if response.data:
            return response.data[0], None

        # Create new chat history
        new_chat = {
            "user_id": user_id,
            "title": "New Chat",
            "messages": [],
            "total_tokens": 0,
            "model_version": "codette-2.0",
            "archived": False,
        }

        response = supabase.table("chat_history").insert(new_chat).execute()
        return response.data[0] if response.data else new_chat, None

    except Exception as e:
        return None, str(e)


async def add_chat_message(
    user_id: str, chat_id: str, role: str, content: str, tokens_used: int = 0
) -> Tuple[Optional[Dict], Optional[str]]:
    """Add message to chat history"""
    if not supabase:
        return None, "Supabase not available"

    try:
        # Insert message
        message = {
            "chat_id": chat_id,
            "role": role,
            "content": content,
            "tokens_used": tokens_used,
        }

        response = supabase.table("chat_message").insert(message).execute()

        # Update chat history total tokens
        if response.data:
            supabase.table("chat_history").update({"total_tokens": "total_tokens + :tokens"}).eq(
                "id", chat_id
            ).execute()

        return response.data[0] if response.data else message, None

    except Exception as e:
        return None, str(e)


async def clear_chat_history(chat_id: str) -> Tuple[bool, Optional[str]]:
    """Clear all messages from chat"""
    if not supabase:
        return False, "Supabase not available"

    try:
        # Delete all messages for this chat
        supabase.table("chat_message").delete().eq("chat_id", chat_id).execute()

        # Reset tokens
        supabase.table("chat_history").update({"total_tokens": 0}).eq("id", chat_id).execute()

        return True, None

    except Exception as e:
        return False, str(e)


# ============================================================================
# MUSIC KNOWLEDGE OPERATIONS
# ============================================================================

async def search_music_knowledge_by_text(
    query: str, category: Optional[str] = None, limit: int = 10
) -> Tuple[Optional[List[Dict]], Optional[str]]:
    """Search music knowledge using full-text search"""
    if not supabase:
        return None, "Supabase not available"

    try:
        # Build query
        q = supabase.table("music_knowledge").select("*")

        if category:
            q = q.eq("category", category)

        # Limit results
        q = q.limit(limit)

        response = q.execute()
        return response.data, None

    except Exception as e:
        return None, str(e)


async def search_music_knowledge_by_similarity(
    embedding: List[float], limit: int = 5
) -> Tuple[Optional[List[Dict]], Optional[str]]:
    """Search music knowledge using vector similarity"""
    if not supabase:
        return None, "Supabase not available"

    try:
        # Call RPC function for vector similarity search
        response = supabase.rpc(
            "search_music_knowledge",
            {
                "query_embedding": embedding,
                "match_count": limit,
            },
        ).execute()

        return response.data, None

    except Exception as e:
        return None, str(e)


async def add_music_knowledge(
    user_id: str, title: str, content: str, category: str, embedding: Optional[List[float]] = None, tags: Optional[List[str]] = None, is_public: bool = True
) -> Tuple[Optional[Dict], Optional[str]]:
    """Add new music knowledge entry"""
    if not supabase:
        return None, "Supabase not available"

    try:
        record = {
            "user_id": user_id,
            "title": title,
            "content": content,
            "category": category,
            "embedding": embedding,
            "tags": tags or [],
            "is_public": is_public,
        }

        response = supabase.table("music_knowledge").insert(record).execute()
        return response.data[0] if response.data else record, None

    except Exception as e:
        return None, str(e)


# ============================================================================
# USER FEEDBACK OPERATIONS
# ============================================================================

async def submit_user_feedback(
    user_id: str, rating: float, feedback_text: str, category: str = "general"
) -> Tuple[Optional[Dict], Optional[str]]:
    """Submit user feedback"""
    if not supabase:
        return None, "Supabase not available"

    try:
        feedback = {
            "user_id": user_id,
            "rating": max(0, min(5, rating)),  # Clamp to 0-5
            "feedback_text": feedback_text,
            "category": category,
        }

        response = supabase.table("user_feedback").insert(feedback).execute()
        return response.data[0] if response.data else feedback, None

    except Exception as e:
        return None, str(e)


# ============================================================================
# API METRICS OPERATIONS
# ============================================================================

async def log_api_metric(
    endpoint: str,
    method: str,
    response_time_ms: float,
    status_code: int,
    user_id: Optional[str] = None,
    metadata: Optional[Dict[str, Any]] = None,
) -> Tuple[Optional[Dict], Optional[str]]:
    """Log API endpoint metric"""
    if not supabase:
        return None, "Supabase not available"

    try:
        metric = {
            "endpoint": endpoint,
            "method": method,
            "response_time_ms": response_time_ms,
            "status_code": status_code,
            "user_id": user_id,
            "metadata": metadata or {},
        }

        response = supabase.table("api_metric").insert(metric).execute()
        return response.data[0] if response.data else metric, None

    except Exception as e:
        return None, str(e)


async def get_average_response_time(
    endpoint: Optional[str] = None, time_window_hours: int = 24
) -> Tuple[Optional[float], Optional[str]]:
    """Get average API response time"""
    if not supabase:
        return None, "Supabase not available"

    try:
        # This would typically be done with a PostgreSQL function
        # For now, return None - implement as RPC function in Supabase
        return None, "Not yet implemented - requires RPC function"

    except Exception as e:
        return None, str(e)


# ============================================================================
# BENCHMARK OPERATIONS
# ============================================================================

async def record_benchmark_result(
    benchmark_type: str,
    score: float,
    metadata: Optional[Dict[str, Any]] = None,
    environment: str = "production",
    model_version: str = "2.0",
) -> Tuple[Optional[Dict], Optional[str]]:
    """Record benchmark result"""
    if not supabase:
        return None, "Supabase not available"

    try:
        result = {
            "benchmark_type": benchmark_type,
            "score": score,
            "metadata": metadata or {},
            "environment": environment,
            "model_version": model_version,
        }

        response = supabase.table("benchmark_result").insert(result).execute()
        return response.data[0] if response.data else result, None

    except Exception as e:
        return None, str(e)


# ============================================================================
# CODETTE FILE OPERATIONS
# ============================================================================

async def upload_file_metadata(
    user_id: str,
    filename: str,
    file_type: str,
    file_size_bytes: int,
    storage_path: Optional[str] = None,
    metadata: Optional[Dict[str, Any]] = None,
) -> Tuple[Optional[Dict], Optional[str]]:
    """Track file upload in database"""
    if not supabase:
        return None, "Supabase not available"

    try:
        record = {
            "user_id": user_id,
            "filename": filename,
            "file_type": file_type,
            "file_size_bytes": file_size_bytes,
            "storage_path": storage_path,
            "metadata": metadata or {},
        }

        response = supabase.table("codette_file").insert(record).execute()
        return response.data[0] if response.data else record, None

    except Exception as e:
        return None, str(e)


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def get_supabase_client() -> Optional[Client]:
    """Get Supabase client instance"""
    return supabase


def is_supabase_available() -> bool:
    """Check if Supabase is available"""
    return supabase is not None and SUPABASE_AVAILABLE


# ============================================================================
# OPERATION GROUPS (for compatibility with frontend imports)
# ============================================================================

# Create operation group objects for easier organization
class ChatHistoryOps:
    @staticmethod
    async def getOrCreate(user_id: str):
        return await get_or_create_chat_history(user_id)
    
    @staticmethod
    async def addMessage(user_id: str, chat_id: str, role: str, content: str, tokens_used: int = 0):
        return await add_chat_message(user_id, chat_id, role, content, tokens_used)
    
    @staticmethod
    async def clear(chat_id: str):
        return await clear_chat_history(chat_id)

class MusicKnowledgeOps:
    @staticmethod
    async def searchByText(query: str, category: Optional[str] = None, limit: int = 10):
        return await search_music_knowledge_by_text(query, category, limit)
    
    @staticmethod
    async def searchBySimilarity(embedding: list, limit: int = 5):
        return await search_music_knowledge_by_similarity(embedding, limit)
    
    @staticmethod
    async def getByCategory(category: str):
        return await search_music_knowledge_by_text("", category, 10)
    
    @staticmethod
    async def add(user_id: str, title: str, content: str, category: str, embedding=None, tags=None, is_public=True):
        return await add_music_knowledge(user_id, title, content, category, embedding, tags, is_public)

class FeedbackOps:
    @staticmethod
    async def submit(feedback_data):
        if hasattr(feedback_data, 'user_id'):
            return await submit_user_feedback(
                feedback_data.user_id,
                feedback_data.rating,
                feedback_data.feedback_text,
                getattr(feedback_data, 'category', 'general')
            )
        return None, "Invalid feedback data"

# Export operation groups
chatHistoryOps = ChatHistoryOps()
musicKnowledgeOps = MusicKnowledgeOps()
feedbackOps = FeedbackOps()
